scipy/stats/_binomtest.py

```diff
@@ -313,16 +312,73 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
             # special case as shortcut, would also be handled by `else` below
             pval = 1.
         elif k < p * n:
-            i = np.arange(np.ceil(p * n), n+1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = n-_binary_search_for_binom_tst(lambda x1: binom.pmf(x1, n, p),
+                                               d*rerr, np.ceil(p * n), n)+1
             pval = binom.cdf(k, n, p) + binom.sf(n - y, n, p)
         else:
-            i = np.arange(np.floor(p*n) + 1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = _binary_search_for_binom_tst(lambda x1: binom.pmf(x1, n, p),
+                                             d*rerr, 0, np.floor(p * n) + 1,
+                                             True)+1
             pval = binom.cdf(y-1, n, p) + binom.sf(k-1, n, p)
 
         pval = min(1.0, pval)
 
     result = BinomTestResult(k=k, n=n, alternative=alternative,
                              proportion_estimate=k/n, pvalue=pval)
     return result
+
+
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+    """
+    Conducts an implicit binary search on a funcion specified by a.
+    args:
+        a: The function over which to perform binary search. Its values
+            for inputs lo and hi should be in ascending or descending
+            order as required by asc_order.
+        d: The value to search.
+        lo: The lower end of range to search.
+        hi: The higher end of the range to search.
+        asc_order: Weather the values returned by the function are in
+                   ascending or descending order.
+    """
+    while lo < hi:
+        mid = (lo+hi)//2
+        midval = a(mid)
+        if midval < d:
+            if asc_order:
+                lo = mid+1
+            else:
+                hi = mid-1
+        elif midval > d:
+            if asc_order:
+                hi = mid-1
+            else:
+                lo = mid+1
+        else:
+            return mid
+    if a(lo) <= d:
+        return lo
+    else:
+        return lo-(asc_order-0.5)*2
+
+
+def test_binomtest():
+    """
+    Some tests for binomtest. Should print all True when run.
+    """
```

Comment:
Please add this to the test suite - again, they're in `scipy/stats/tests/test_morestats.py` starting at Line 663.
Ideally, you'd [`@parametrize`](https://docs.pytest.org/en/stable/parametrize.html) the test, and you should use `assert_allclose` to compare the values.